cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/concurrency/ExecutionConcurrencyManager.java

```diff
@@ -0,0 +1,271 @@
+    if (requestedMaxInterBrokerPartitionMovements != null
+        && requestedMaxInterBrokerPartitionMovements > _maxNumClusterMovementConcurrency) {
+      throw new IllegalArgumentException("Attempt to set max inter-broker partition movements [" + requestedMaxInterBrokerPartitionMovements
+                                         + "] to greater than the maximum" + " number of allowed movements in cluster ["
+                                         + _maxNumClusterMovementConcurrency + "].");
+    }
+    _requestedMaxInterBrokerPartitionMovements = requestedMaxInterBrokerPartitionMovements;
+  }
+
+  /**
+   * @return Allowed upper bound of inter broker partition movements in cluster
+   */
+  public int maxInterBrokerPartitionMovements() {
+    return _requestedMaxInterBrokerPartitionMovements == null ? _defaultMaxInterBrokerPartitionMovements
+                                                              : _requestedMaxInterBrokerPartitionMovements;
+  }
+
+  /**
+   * Get the concurrency of each broker without per-broker throttling. This concurrency is simply the overall allowed concurrency divided
+   *  by broker count.
+   * @param brokersWithReplicaMoves The set of brokers that involve in replica move
+   * @param brokersToSkipConcurrencyCheck The set of brokers that is configured to skip concurrency check.
+   * @return concurrency of each broker without per-broker throttling
+   */
+  public int unthrottledConcurrency(Set<Integer> brokersWithReplicaMoves, Set<Integer> brokersToSkipConcurrencyCheck) {
+    int numUnthrottledBrokers = (int) brokersWithReplicaMoves.stream().filter(brokersToSkipConcurrencyCheck::contains).count();
+    if (numUnthrottledBrokers == 0) {
+      // All brokers are throttled.
+      return Integer.MAX_VALUE;
+    }
+    int unthrottledConcurrency = _maxNumClusterMovementConcurrency / numUnthrottledBrokers;
+    LOG.debug("Unthrottled concurrency is {} for {} brokers.", unthrottledConcurrency, numUnthrottledBrokers);
+    return unthrottledConcurrency;
+  }
+
+  /**
+   * Get execution the concurrency summary that can show the avg/min/max allowed broker concurrency
+   * @return the execution concurrency summary
+   */
+  public synchronized ExecutionConcurrencySummary getExecutionConcurrencySummary() {
+    if (!_initialized) {
+      throw new IllegalArgumentException("The execution manager has not been initialized.");
+    }
+    return _executionConcurrencySummary;
+  }
+
+  private void sanityCheckRequestedConcurrency(Integer concurrency, ConcurrencyType concurrencyType) {
+    if (concurrency != null && concurrency >= _maxNumClusterMovementConcurrency) {
+      throw new IllegalArgumentException("Attempt to set " + concurrencyType + " concurrency ["
+                                         + concurrency + "] to greater than or equal to the maximum"
+                                         + " number of allowed movements in cluster [" + _maxNumClusterMovementConcurrency + "].");
```

Comment:
Same thing here. From what I've seen, even our SREs (who use CC every single day) struggle to understand the complexity and how various configs and goals interact with each other. There should be a conscious effort to simplify the operation of CC itself wherever possible, and that means not just logging sanity-check failures but actively correcting them where possible and simultaneously instructing users on how to fix things if the automatic correction is not what they wanted. (One might go so far as to include a new metric indicating the number of sanity-check auto-adjustments since startup so the potential user error isn't just buried in a log file. Or it could be included in the cluster-state command-line report or a new ccdo "things you should be aware of" option. Or both.)